doc: Document the JSON logging functionality in the proxy (#3667)
kroxylicious/kroxylicious: kroxylicious-docs/docs/_modules/monitoring/proc-proxy-setting-log-levels.adoc

Reviewer: This module mixes concept and procedure content, so it is not fully modular.

Suggest splitting into:

* a concept module for logging configuration (environment variables, setting config)
* Procedure module (for example, switching to JSON format)

To keep this PR focused on the new additions, we can address this in a follow-up PR rather than here.

I’m happy to take this on once this PR is merged, if that helps.

@robobario @tombentley

Reviewer 2: Thanks @PaulRMellor, SGTM

Reviewer: Change:

```
.Switching back to text format

Remove the environment variable or set it to `text`:
Remove the environment variable or set it explicitly to `text`:

[source,bash]
----
unset KROXYLICIOUS_LOG_FORMAT
# or
export KROXYLICIOUS_LOG_FORMAT=text
----
```

To:

```
. To switch back to text format, remove the environment variable or set it to `text`:
+
[source,bash]
----
unset KROXYLICIOUS_LOG_FORMAT
# or
export KROXYLICIOUS_LOG_FORMAT=text
----
```

Reviewer: This gives us two steps in the procedure to make it clearer and more task-based.

The Kroxylicious {github-release}[binary distribution^] includes https://logging.apache.org/log4j/2.x[log4j2] as the default logging backend.

When using the `bin/kroxylicious-start.sh` script from the binary distribution, you can configure logging behaviour using the following environment variables.

.Logging environment variables
[cols="2,1,3",options="header"]
|===
|Environment variable |Default |Description

|`KROXYLICIOUS_ROOT_LOG_LEVEL`
|`WARN`

> |Log level for all loggers not otherwise configured.

Reviewer 2: We should document what the valid values are, in addition to the default.

|(built-in defaults)
|Overrides the entire set of log4j2 JVM system properties. When set, the default `-Dlog4j2.configurationFile` and `-Dlog4j2.contextSelector` options are replaced, so you must include all required options explicitly.
|===

.Setting the application log level to DEBUG
[source,properties]
----
KROXYLICIOUS_APP_LOG_LEVEL="DEBUG"
----

NOTE: Setting the log level to `DEBUG` or `TRACE` will produce very verbose logs.

.Loading a custom log4j2 configuration file
[source,bash]
----
export KROXYLICIOUS_LOGGING_OPTIONS="-Dlog4j2.configurationFile=/path/to/custom/log4j2.yaml \
  -Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector"
----

WARNING: Setting `KROXYLICIOUS_LOGGING_OPTIONS` replaces the built-in defaults entirely.
Always include `-Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector`
to preserve asynchronous logging.

[id='proc-proxy-json-log-format-{context}']
== Switching to JSON log format

By default, Kroxylicious outputs logs in human-readable text format.
You can switch to structured JSON output for integration with log aggregation systems such as the Elastic Stack (ELK) or Splunk.

> The JSON format follows the https://www.elastic.co/guide/en/ecs/current/index.html[Logstash JSON Event Layout V1] schema.

Reviewer 3: The link is mismatched, we do use Logstash JSON Event Layout V1 but it links to ecs docs

Maybe we should link to https://logging.apache.org/log4j/2.x/manual/json-template-layout.html and say something like "uses Log4j2's JsonTemplateLayout with the Logstash JSON Event V1 format"

The Kroxylicious {github-release}[binary distribution^] includes https://logging.apache.org/log4j/2.x[log4j2] as the default logging backend.

> When using the `bin/kroxylicious-start.sh` script from the binary distribution, you can configure logging behaviour using the following environment variables.

Reviewer: ```suggestion
When you use the `bin/kroxylicious-start.sh` script from the binary distribution, you can configure logging behavior using the following environment variables.
```

.Logging environment variables
[cols="2,1,2,3",options="header"]
|===
|Environment variable |Default |Valid values |Description

|`KROXYLICIOUS_ROOT_LOG_LEVEL`
|`WARN`
|`TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR`, `FATAL`, `OFF`

> |Log level for all loggers not otherwise configured.

Reviewer: ```suggestion
|Log level for all loggers that are not otherwise configured.
```

|`KROXYLICIOUS_APP_LOG_LEVEL`
|`INFO`
|`TRACE`, `DEBUG`, `INFO`, `WARN`, `ERROR`, `FATAL`, `OFF`

> |Log level for Kroxylicious loggers (`io.kroxylicious.*`). Takes precedence over `KROXYLICIOUS_ROOT_LOG_LEVEL` for application code.

Reviewer: ```suggestion
|Log level for Kroxylicious loggers (`io.kroxylicious.*`). This setting takes precedence over `KROXYLICIOUS_ROOT_LOG_LEVEL` for application code.
```

|`KROXYLICIOUS_LOG_FORMAT`
|`text`
|`text`, `json`

> |Output format for log messages. Use `json` to enable structured JSON output.

Reviewer: ```suggestion
|Output format for log messages. Set to `json` to enable structured JSON output.
```

|`KROXYLICIOUS_LOG_JSON_TEMPLATE`
|`classpath:LogstashJsonEventLayoutV1.json`

> |Any valid https://logging.apache.org/log4j/2.x/manual/json-template-layout.html[Log4j2 JsonTemplateLayout] event template URI

Reviewer: ```suggestion
|Any valid https://logging.apache.org/log4j/2.x/manual/json-template-layout.html[Log4j2 JsonTemplateLayout] event template URI.
```

|JSON layout template used when `KROXYLICIOUS_LOG_FORMAT=json`. For example, use `classpath:EcsLayout.json` for Elastic Common Schema or `classpath:GelfLayout.json` for GELF format.

|`KROXYLICIOUS_LOGGING_OPTIONS`
|(built-in defaults)

> |Any valid log4j2 JVM system properties

Reviewer: ```suggestion
|Any valid log4j2 JVM system properties.
```

> |Overrides the entire set of log4j2 JVM system properties. When set, the default `-Dlog4j2.configurationFile` and `-Dlog4j2.contextSelector` options are replaced, so you must include all required options explicitly.

Reviewer: ```suggestion
|Overrides the entire set of log4j2 JVM system properties. When set, the default `-Dlog4j2.configurationFile` and `-Dlog4j2.contextSelector` options are replaced. You must include all required options explicitly.
```

|===

> .Setting the application log level to DEBUG

Reviewer: ```suggestion
The following examples show how to set the application log level and configure a custom Log4j2 configuration.

.Setting the application log level to DEBUG
```

[source,properties]
----
KROXYLICIOUS_APP_LOG_LEVEL="DEBUG"
----

> NOTE: Setting the log level to `DEBUG` or `TRACE` will produce very verbose logs.

Reviewer: ```suggestion
NOTE: Setting the log level to `DEBUG` or `TRACE` produces very verbose logs.
```

.Loading a custom log4j2 configuration file
[source,bash]
----
export KROXYLICIOUS_LOGGING_OPTIONS="-Dlog4j2.configurationFile=/path/to/custom/log4j2.yaml \
  -Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector"
----

> WARNING: Setting `KROXYLICIOUS_LOGGING_OPTIONS` replaces the built-in defaults entirely.

Reviewer: ```suggestion
WARNING: Setting `KROXYLICIOUS_LOGGING_OPTIONS` replaces the built-in defaults.
```

Always include `-Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector`

> to preserve asynchronous logging.

Reviewer: ```suggestion
Always include `-Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector` to preserve asynchronous logging.
```

[id='proc-proxy-json-log-format-{context}']
== Switching to JSON log format

By default, Kroxylicious outputs logs in human-readable text format.

> You can switch to structured JSON output for integration with log aggregation systems such as the Elastic Stack (ELK) or Splunk.

Reviewer: ```suggestion
You can switch to structured JSON output to integrate with log aggregation systems such as the Elastic Stack (ELK) or Splunk.
```

The JSON format uses https://logging.apache.org/log4j/2.x/manual/json-template-layout.html[Log4j2's JsonTemplateLayout] with the Logstash JSON Event V1 format.

> Structured context fields emitted by Kroxylicious appear as top-level fields alongside the standard log metadata.

Reviewer: ```suggestion
Structured context fields emitted by Kroxylicious appear as top-level fields alongside standard log metadata.
```

.Procedure

. Set the `KROXYLICIOUS_LOG_FORMAT` environment variable before starting the proxy:
+
[source,bash]
----
export KROXYLICIOUS_LOG_FORMAT=json
bin/kroxylicious-start.sh
----

> .Example JSON log output

Reviewer: ```suggestion
+
Example JSON log output:
```

[source,bash]
----
export KROXYLICIOUS_LOGGING_OPTIONS="-Dlog4j2.configurationFile=/path/to/custom/log4j2.yaml \
  -Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector"
----

WARNING: Setting `KROXYLICIOUS_LOGGING_OPTIONS` replaces the built-in defaults.
Always include `-Dlog4j2.contextSelector=org.apache.logging.log4j.core.async.AsyncLoggerContextSelector` to preserve asynchronous logging.

[id='proc-proxy-json-log-format-{context}']
== Switching to JSON log format

By default, Kroxylicious outputs logs in human-readable text format.
You can switch to structured JSON output to integrate with log aggregation systems such as the Elastic Stack (ELK) or Splunk.

The JSON format uses https://logging.apache.org/log4j/2.x/manual/json-template-layout.html[Log4j2's JsonTemplateLayout] with the Logstash JSON Event V1 format.
Structured context fields emitted by Kroxylicious appear as top-level fields alongside standard log metadata.

.Procedure

. Set the `KROXYLICIOUS_LOG_FORMAT` environment variable before starting the proxy:
+
[source,bash]
----
export KROXYLICIOUS_LOG_FORMAT=json
bin/kroxylicious-start.sh
----

> Example JSON log output:

Reviewer: ```suggestion
+
Example JSON log output:
+
```

  "thread_name": "kafka-request-handler-1",
  "level": "INFO",
  "level_value": 20000,
  "sessionId": "abc-123",
  "clientId": "producer-1",
  "apiKey": "Produce"
}
----

When an exception is logged, a `stack_trace` field is also included in the output.

.Switching back to text format

Remove the environment variable or set it explicitly to `text`:

[source,bash]
----
unset KROXYLICIOUS_LOG_FORMAT
# or
export KROXYLICIOUS_LOG_FORMAT=text
----

[id='con-proxy-structured-log-fields-{context}']

> == Structured log fields

Reviewer 3: this is part of a `proc` file which covers a single procedure, maybe this section describing the `Structured log fields` concepts should be extracted into a `con-` file and referenced if required.

  "thread_name": "kafka-request-handler-1",
  "level": "INFO",
  "level_value": 20000,
  "sessionId": "abc-123",
  "clientId": "producer-1",
  "apiKey": "Produce"
}
----

> When an exception is logged, a `stack_trace` field is also included in the output.

Reviewer: ```suggestion
+
When an exception is logged, a `stack_trace` field is included in the output.
```

  "thread_name": "kafka-request-handler-1",
  "level": "INFO",
  "level_value": 20000,
  "sessionId": "abc-123",
  "clientId": "producer-1",
  "apiKey": "Produce"
}
----

> When an exception is logged, a `stack_trace` field is included in the output.

Reviewer: ```suggestion
+
When an exception is logged, a `stack_trace` field is included in the output.
```

> . To switch back to text format, remove the environment variable or set it to `text`:

Reviewer: ```suggestion
. To switch back to text format, remove the environment variable or set it to `text`:
+
```
